config/src/test/java/org/springframework/security/config/annotation/web/configurers/AnonymousConfigurerTests.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.config.annotation.web.configurers;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.config.test.SpringTestRule;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Controller;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+
+/**
+ * @author Rob Winch
+ * @author Josh Cummings
+ */
+public class AnonymousConfigurerTests {
+	@Rule
+	public final SpringTestRule spring = new SpringTestRule();
+
+	@Autowired
+	private MockMvc mockMvc;
+
+	@Test
+	public void requestWhenAnonymousTwiceInvokedThenDoesntOverride() throws Exception {
+		this.spring.register(InvokeTwiceDoesNotOverride.class).autowire();
+
+		MvcResult result = this.mockMvc.perform(get("/")).andReturn();
+
+		assertThat(result.getResponse().getContentAsString()).isEqualTo("custom");
+	}
+
+	@EnableWebSecurity
+	static class InvokeTwiceDoesNotOverride extends WebSecurityConfigurerAdapter {
+
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			http
+				.anonymous()
+				.principal("custom")
```

Comment:
The groovy test was configuring `key` rather than `principal`

Reply:
Checking the key has the downside that it gets hashed by the `AnonymousAuthenticationToken` constructor, meaning that to confirm it behaviorally, I'd need to exercise the same hashing logic in the test, e.g.:

```assertThat(result.getResponse().getContentAsString()).isEqualTo("custom".hashCode());```
which is why I changed it to principal. This wasn't a problem in the original test since it just verified the structure of `AnonymousAuthenticationFilter`.

I can test the key for sure; I'm just duplicating some logic that is internal to `AnonymousAuthenticationToken`. And since both key and principal are configurable values on `AnonymousAuthenticationFilter`, it seemed like a fair trade-off for confirming that configs aren't overridden.

Are there other things I should consider here? (Other than the realization that I should have put this explanation in my original commit to explain why I departed from the original test :) )

Reply:
If you pass in the same `key` to `AnonymousAuthenticationToken` as is configured via `.anonymous().key("custom")` than the test should pass as the hash check will succeed in the `AnonymousAuthenticationProvider`. So try this in your test and it should work:

```
AnonymousAuthenticationToken authRequest = new AnonymousAuthenticationToken("custom", "anonymous", AuthorityUtils.createAuthorityList("ROLE_ANONYMOUS"));
		
MvcResult result = this.mockMvc.perform(get("/").with(authentication(authRequest))).andReturn();
```